CRM/Core/EntityTokens.php

```diff
@@ -16,7 +16,14 @@
 use Civi\Token\TokenProcessor;
 
 /**
- * Class CRM_Core_EntityTokens
+ * Generic base-class which loads tokens via APIv4.
+ *
+ * To write a subclass:
+ *
+ * - (MUST) Implement getEntityName() and getApiEntityName()
+ * - (MUST) Implement getApiTokens()
+ * - (MAY) Implement getAliasTokens()
```

Comment:
Lets call it `getDeprecatedTokens` - we want to have one preferred token for any output - not some loosely goosey 'any one will do' - tbh - I would have left adding either until we need them which we shouldn't for this PR

Reply:
"Deprecated" describes the policy or preference - and it means "Show warnings to admins!". "Alias" describes the functionality or runtime-execution - and it means "Substitute the value at runtime!" These are orthogonal dimensions.

This needs the alias to pass `CRM_Contribute_ActionMapping_ByTypeTest` (which favors the new/not-really-supported `campaign_id:label` over the existing `campaign_id.title`), and the callers which consume `getAliasTokens()` are mostly using it to provide runtime-substitution behavior.

A way to deprecate tokens sounds useful, but we don't really need it for adding support for `contributionId` or for adding support for true-APIv4-loading. Deprecation is orthogonal. (Ex: In this case, if I recall your intent correctly, `campaign_id:label` is upcoming - that's why it was added to the unit-test. The alias here is *in anticipation of an upcoming canonical form.* That's not deprecation.) Moreover, doing actual deprecation would mean a different set of tasks - eg warnings/flags in various admin screens.

Reply:
getLegacyTokens maybe? ( We don't actually need it for this PR as we don't have any in the contribution entity)

Reply:
Actually - not deprecated implies - at some point in the future we will do something to let admins know not to use these.

Note campaign_id.title isn't existing - this PR would potentially make it work as a hidden & not preferred token.

campaign_id:label currently works in already-merged prs & this PR doesn't need to do anything specific about it

Reply:
> campaign_id:label currently works in already-merged prs & this PR doesn't need to do anything specific about it

Aah, cool. I missed that merge. Just double-checked/confirmed that APIv4 now reads `campaign_id:label`.

> Note campaign_id.title isn't existing - campaign_id:label exists and works already

Right, from token POV. From APIv4 POV, the sequencing was the opposite. Because why should life be easy? ;)

> there are no aliases or deprecated tokens that are relevant to the contribution token entity any more

I'll update the references to `campaign_id` derivations.

Note that  `{contribution.campaign}` still hangs over us. It's 1 liner as an alias.

Reply:
@totten we don't actually need to handle the deprecated contribution.campaign in the token processor - it's only even been in the legacy code. We can deprecate it out of there more actively at some point

Reply:
@totten I've updated https://github.com/civicrm/civicrm-core/pull/21134 so that it actually passes & adds in the same listening that this does for clarity.  It doesn't incorporate more recent changes about preferred interaction with the action schedule processor

Reply:
@eileenmcnaughton @totten `getAliasTokens()` makes the most technical sense here to me. It's what I chose when doing the same thing and describes technically *exactly* what it does - which is map one token name to another without any processing. Both *deprecated* and *legacy* tokens implies that they might be processed differently which they are not.

It does not prevent us following a process later to actually deprecate/remove them and I am all for simplifying what tokens are allowed in the future.

I see the concept has gone in API4 (good) but it was pretty common in API3:

![image](https://user-images.githubusercontent.com/2052161/129548577-8413f034-cffa-42f4-a390-69b74a3e159a.png)

Reply:
yeah - wanting to not go back to the apiv3 approach has definitely been part of my thinking. It was something we did because of what we were starting with but I would definitely have more clearly highlighted them as deprecated rather than aliases if I were doing apiv3 now (I was the person who added aliases to apiv3)